### PR TITLE
[CodeCompletion] Batch code completion test 

### PR DESCRIPTION
**Pull request by rintaro**
Utilize fast completion (reusing `CompilerInstance`) in libIDE code completion tests.

- Catch regressions in fast-completion
- Reduce the test execution time (e.g. 32sec -> approx. 2sec  for `complete_types.swift`)

Introduce `-batch-code-completion` mode. The typical usage is:
```
// RUN: %swift-ide-test -batch-code-completion \
// RUN:     -filecheck %raw-FileCheck \
// RUN:     -completion-output-dir %t \
// RUN:     -source-filename %s
```

This scans the source file for tokens (`#^NAME^#`), then performs code completion for each token in *random* order with `reuseastcontext` mode.

Options are:
* `-filecheck`: `FileCheck` command path. Completion results are checked with this `FileCheck`
* `-completion-output-dir`: Temporary file directory. Completion results are output with `${output dir}/complete-${token name}.result`
* `-random-seed`: By default, tokens are shuffled with random generated seed. Using `-random-seed` you can specify shuffle seed. This is convenient for reproducing order-dependent failures
* `-code-completion-token`: You can specify a specific token name to test, but in `batch-code-completion` mode

Each token can have several options (e.g. `#^NAME?check=GLOBAL;comments=true^#`):
* `check`: By default, the completion results are `FileCheck -check-prefixes=`ed with the token name. `check` can changes the `check-prefix` name. Also, you can specify multiple `check` to perform multiple `FileCheck`s with different prefixes
* `keywords`: equivalent to `-code-completion-keywords`
* `comments`: equivalent to `-code-completion-comments`
* `skip`: Tokens with `skip` are skipped

In this PR. Only migrate `test/IDE/complete_type.swift`. But I intend to migrate other test files in follow up PRs.

rdar://problem/16025703

### CODE REVIEW

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

File: test/IDE/complete_type.swift [hunk 1]
```diff
@@ -698,173 +298,173 @@ func resyncParser1() {}
...
 }
 
 func testTypeInTupleType6() {
-  var localVar: (a:, #^TYPE_IN_TUPLE_TYPE_6^#
+  var localVar: (a:, #^TYPE_IN_TUPLE_TYPE_6?check=WITH_GLOBAL_TYPES;check=GLOBAL_NEGATIVE^#
 }
 
 func testTypeInTupleType7() {
-  var localVar: (a: b: #^TYPE_IN_TUPLE_TYPE_7^#
+  var localVar: (a: b: #^TYPE_IN_TUPLE_TYPE_7?skip=FIXME^#
```
Comment (rintaro) [inline test/IDE/complete_type.swift:410]: This was not checked. i.e. no `RUN: ` line was there.

File: test/IDE/complete_type.swift [hunk 2]
```diff
@@ -1148,8 +748,8 @@ extension Task.Inner {
     fatalError()
   }
 }
-extension Task.Inner.#^UNBOUND_DOT^# {}
-func testUnbound(x: Task.Inner.#^UNBOUND_DOT^#) {}
+extension Task.Inner.#^UNBOUND_DOT_1?check=UNBOUND_DOT^# {}
+func testUnbound(x: Task.Inner.#^UNBOUND_DOT_2?check=UNBOUND_DOT^#) {}
```
Comment (rintaro) [inline test/IDE/complete_type.swift:752]: `UNBOUND_DOT` used to be duplicated.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 3]
```diff
@@ -876,6 +905,392 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+                              "complete-" + Token.Name + ".result");
+      if (auto res = llvm::sys::fs::openFileForWrite(resultFilename, resultFD)) {
+        llvm::errs() << "error: failed to create output file: "
+                     << resultFilename << "\n";
+        return res.value();
+      }
+      shouldClose = true;
+    } else {
+      // If output directory is not specified, print the results to STDOUT.
+      resultFD = llvm::sys::fs::getStdoutHandle();
```
Comment (rintaro) [inline tools/swift-ide-test/swift-ide-test.cpp:1197]: https://ci-external.swift.org/job/swift-PR-windows/4317/
```
error C2440: '=': cannot convert from 'llvm::sys::fs::file_t' to 'int'
```
Will fix

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 4]
```diff
@@ -403,6 +409,29 @@ DebugForbidTypecheckPrefix("debug-forbid-typecheck-prefix",
...
+                    llvm::cl::init(0));
+
+
+static llvm::cl::opt<std::string>
+CompletionOutputDir("completion-output-dir", llvm::cl::value_desc("path"),
+                    llvm::cl::desc("Directory for completion output"),
+                    llvm::cl::cat(Category));
+
+static llvm::cl::opt<std::string>
+FileCheckPath("use-filecheck", llvm::cl::value_desc("path"),
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:427]: I suggest just "--filecheck"

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 5]
```diff
@@ -403,6 +409,29 @@ DebugForbidTypecheckPrefix("debug-forbid-typecheck-prefix",
...
+                               llvm::cl::desc("Path to 'FileCheck' utility"),
+                               llvm::cl::cat(Category));
+
+static llvm::cl::opt<bool>
+DontFileCheck("dont-filecheck", llvm::cl::desc("Skip 'FileCheck' checking"),
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:432]: I suggest "--skip-filecheck", which is more consistent with other options.

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 6]
```diff
@@ -3526,6 +3940,18 @@ int main(int argc, char *argv[]) {
   case ActionType::DumpCompletionCache:
     llvm_unreachable("should be handled above");
 
+  case ActionType::BatchCodeCompletion:
+    if (options::FileCheckPath.empty()) {
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:3944]: `&& !options::DontFileCheck`

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 7]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+      return true;
+    }
+
+    InputPtr = Ptr + 2;
+    return false;;
+  }
+};
+
+static std::unique_ptr<llvm::MemoryBuffer>
+removeCodeCompletionTokens(llvm::MemoryBuffer *Input,
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1017]: What do you think of migrating the one-shot completion path to use this too?  Can we centralize on one implementation?  We could always make it an error if the query string is used in a way we don't support.

This could be a future change, not necessarily for this PR.

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 8]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+  auto FileBufOrErr = llvm::MemoryBuffer::getFile(SourceFilename);
+  if (!FileBufOrErr) {
+    llvm::errs() << "error opening input file: "
+                 << FileBufOrErr.getError().message() << '\n';
+    return 1;
+  }
+
+  std::string Error;
+
+  llvm::SmallVector<CompletionTestToken, 0> CCTokens;
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1090]: Should we just change to `std::vector` if we're not using small storage?

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 9]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+    // Shuffle tokens to detect order-dependent bugs.
+    if (CCTokens.empty()) {
+      llvm::errs()
+          << "error: could not find any code completion tokens in input file\n";
+      return 1;
+    }
+    unsigned RandomSeed = options::RandomSeed;
+    if (RandomSeed == 0)
+      RandomSeed = std::chrono::system_clock::now().time_since_epoch().count();
+    llvm::errs() << "Shuffle with --random-seed=" << RandomSeed << "\n";
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1126]: I suggest rewording this:

Use --random-seed= to reproduce the order of this run.

I think this makes it clearer what is happening.  Using "Shuffle" in the imperative form makes it sound like instructions to the user, but this is only relevant if you are trying to reproduce something.

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 10]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+
+    std::shuffle(CCTokens.begin(), CCTokens.end(),
+                 std::default_random_engine(RandomSeed));
+  }
+
+  // Compleiton results are output to
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1132]: Typo s/Compleiton/Completion/

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 11]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+          auto *SF = CI.getCodeCompletionFile();
+          performCodeCompletionSecondPass(*SF, *callbacksFactory);
+        });
+    auto completionEnd = std::chrono::high_resolution_clock::now();
+    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+        completionEnd - completionStart);
+    llvm::errs() << "Elapsed: " << elapsed.count() << " msec\n";
+    OS.flush();
+
+    if (OutputDir.empty()) {
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1218]: I was surprised that empty output dir implies skipping filecheck.  Should we make it an error to have no output dir unless you also have --dont-filecheck ?

File: tools/swift-ide-test/swift-ide-test.cpp [hunk 12]
```diff
@@ -876,6 +905,391 @@ static int doCodeCompletion(const CompilerInvocation &InitInvok,
...
+      llvm::errs() << "error: " << Error << "\n";
+      return 1;
+    }
+
+    if (options::FileCheckPath.empty() || options::DontFileCheck)
+      continue;
+
+    bool isFileCheckFailed = false;
+    for (auto Prefix : Token.CheckPrefixes) {
+      SmallVector<StringRef, 6> FileCheckArgs = {
```
Comment (benlangmuir) [inline tools/swift-ide-test/swift-ide-test.cpp:1249]: This is just an array, does it really need to be a vector?

**Review by benlangmuir — COMMENTED** [review]
Really happy you're working on this.  It will make a huge difference in test performance once we port all the big tests.

File: tools/swift-ide-test/swift-ide-test.cpp (hunk 7 above)
Comment (rintaro) [reply tools/swift-ide-test/swift-ide-test.cpp:1017]: Yeah, I also am thinking about it. Will definitely do it in future.

File: tools/swift-ide-test/swift-ide-test.cpp (hunk 8 above)
Comment (rintaro) [reply tools/swift-ide-test/swift-ide-test.cpp:1090]: LLVM programmer's manual actually encourages to use `SmallVector<T, 0>` over `std::vector<T>`.
https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

> SmallVector has grown a few other minor advantages over std::vector, causing SmallVector<Type, 0> to be preferred over std::vector<Type>.
> 
> 1. std::vector is exception-safe, and some implementations have pessimizations that copy elements when SmallVector would move them.
> 2. SmallVector understands llvm::is_trivially_copyable<Type> and uses realloc aggressively.
> 3. Many LLVM APIs take a SmallVectorImpl as an out parameter (see the note below).
> 4. SmallVector with N equal to 0 is smaller than std::vector on 64-bit platforms, since it uses unsigned (instead of void*) for its size and capacity.

Comment (benlangmuir) [reply tools/swift-ide-test/swift-ide-test.cpp:1090]: Okay, if that's the new guidance I'm fine with doing this for consistency with other code.  I will point out that (2) and (3) don't apply at all here and (1) and (4) don't matter for this use case.

File: tools/swift-ide-test/swift-ide-test.cpp (hunk 11 above)
Comment (rintaro) [reply tools/swift-ide-test/swift-ide-test.cpp:1223]: Not specifying '-completion-output-dir' is now error unless `-skip-filecheck`

File: tools/swift-ide-test/swift-ide-test.cpp (hunk 12 above)
Comment (rintaro) [reply tools/swift-ide-test/swift-ide-test.cpp]: No. Updated.

**Review by benlangmuir — APPROVED** [review]

**Review by akyrtzi — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Windows: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-windows/4334/
